webhook_redelivery.py

```diff
@@ -0,0 +1,412 @@
+#  Copyright 2024 Canonical Ltd.
+#  See LICENSE file for licensing details.
+
+"""Redeliver failed webhooks since a given time.
+
+Only webhooks with action type queued are redelivered (as the others are not routable).
+"""
+import argparse
+import json
+import logging
+import os
+import sys
+from collections import namedtuple
+from dataclasses import dataclass
+from datetime import datetime, timedelta, timezone
+from functools import wraps
+from typing import Callable, Iterator, ParamSpec, TypeVar
+
+from github import BadCredentialsException, Github, GithubException, RateLimitExceededException
+from github.Auth import AppAuth, AppInstallationAuth, Token
+from pydantic import BaseModel
+
+from webhook_router.app import SUPPORTED_GITHUB_EVENT
+from webhook_router.router import ROUTABLE_JOB_STATUS
+
+ARG_PARSE_ERROR_EXIT_CODE = 1
+REDELIVERY_ERROR_EXIT_CODE = 2
+
+GITHUB_TOKEN_ENV_NAME = "GITHUB_TOKEN"
+GITHUB_APP_CLIENT_ID_ENV_NAME = "GITHUB_APP_CLIENT_ID"
+GITHUB_APP_INSTALLATION_ID_ENV_NAME = "GITHUB_APP_INSTALLATION_ID"
+GITHUB_APP_PRIVATE_KEY_ENV_NAME = "GITHUB_APP_PRIVATE_KEY"
+
+OK_STATUS = "OK"
+
+P = ParamSpec("P")
+R = TypeVar("R")
+
+logger = logging.getLogger(__name__)
+
+
+class GithubAppAuthDetails(BaseModel):
+    """The details to authenticate with Github using a Github App.
+
+    Attributes:
+        client_id: The Github App client ID.
+        installation_id: The installation ID of the Github App.
+        private_key: The private key to authenticate with Github.
+    """
+
+    client_id: str
+    installation_id: int
+    private_key: str
+
+
+GithubToken = str
+GithubAuthDetails = GithubAppAuthDetails | GithubToken
+
+_ParsedArgs = namedtuple("_ParsedArgs", ["since", "github_auth_details", "webhook_address"])
+
+
+@dataclass
+class WebhookAddress:
+    """The address details to identify the webhook.
+
+    Attributes:
+        github_org: Github organisation where the webhook is registered.
+        github_repo: Github repository, where the webhook is registered. Only applicable for
+            repository webhooks.
+        id: The identifier of the webhook.
+    """
+
+    github_org: str
+    github_repo: str | None
+    id: int
+
+
+@dataclass
+class _WebhookDeliveryAttempts:
```

Comment:
Should this class be multiple (i.e., end with s) or be `_WebhookDeliveryAttempt`?